lib/swig/GeNNModel.py

```diff
@@ -0,0 +1,934 @@
+        if csName in self.currentSources:
+            raise ValueError( 'current source "{0}" already exists'.format( csName ) )
+
+        cSource = CurrentSource( csName )
+        cSource.setCurrentSourceModel( currentSourceModel, paramSpace, varSpace )
+        cSource.addTo( self._model, self.neuronPopulations[popName] )
+
+        self.currentSources[csName] = cSource
+
+        return cSource
+
+
+    def setConnections( self, popName, conns, g ):
+        self.synapsePopulations[popName].setConnections( conns, g )
+
+
+    def initializeVarOnDevice( self, popName, varName, mask, vals ):
+        """Set values for the given variable and population and push them to the device
+        
+        Note: shapes of mask and vals must be the same
+
+        Args:
+        popName -- name of the population for which the values must be set
+        varName -- name of the variable for which the values must be set
+        mask    -- list with neurons ids
+        vals    -- list with variable values
+        """
+        if popName not in self.neuronPopulations:
+            if popName not in self.synapsePopulations:
+                raise ValueError( 'Failed to initialize variable "{0}": \
+                        population "{1}" does not exist'.format( varName, popName ) )
+            else:
+                var = self.synapsePopulations[popName].vars[varName]
+        else:
+            var = self.neuronPopulations[popName].vars[varName]
+
+        var.view[mask] = vals
+        self.pushStateToDevice( popName )
+
+
+    def initializeSpikesOnDevice( self, popName, mask, targets, counts ):
+        """Set spike counts and targets for the given population and push them to the device
+        
+        Note: shapes of mask, targets and counts must be the same
+
+        Args:
+        popName -- name of the population for which the spikes must be set
+        mask    -- list with source neurons ids
+        targets -- list with target neuron ids
+        counts  -- list with number of spikes for source neurons
+        """
+        if popName not in self.neuronPopulations:
+            raise ValueError( 'Failed to initialize variable "{0}": \
+                    population "{1}" does not exist'.format( varName, popName ) )
+        self.neuronPopulations[popName].spikes[mask] = targets
+        self.neuronPopulations[popName].spikeCount[mask] = counts
+        self.pushSpikesToDevice( popName )
+
+
+    def build( self, pathToModel = "./" ):
+
+        """Finalize and build a GeNN model
+        
+        Keyword args:
+        pathToModel -- path where to place the generated model code. Defaults to the local directory.
+        """
+
+        if self._built:
+            raise Exception("GeNN model already built")
+        self._pathToModel = pathToModel
+
+        for popName, popData in self.synapsePopulations.items():
+            if popData.sparse:
+                popData.pop.setMaxConnections( popData.size )
```

Comment:
For sparse projections, max connections means (apologies it is badly named) the largest number of postsynaptic neurons any presynaptic neuron is connected to i.e. the longest length of a sparse synaptic matrix row - on that basis I'm not sure if this is correct.

Reply:
oh. I clearly misunderstood the name.